Implement aggregate_metric field mapper

csoulios opened the pull request:
This PR addresses part of #42720 creating a field mapper for storing pre-computed aggregate metrics such as `min`, `max`, `sum` and `value_count`

csoulios committed:
Implemented aggregate_metric field type for storing pre-computed aggregates

elasticmachine commented:
Pinging @elastic/es-analytics-geo (:Analytics/Rollup)

iverase commented on x-pack/plugin/rollup/src/main/java/org/elasticsearch/xpack/rollup/mapper/AggregateMetricFieldMapper.java:
```diff
@@ -0,0 +1,340 @@
+                                                                         Map<String, Object> node,
+                                                                         ParserContext parserContext) throws MapperParsingException {
+            AggregateMetricFieldMapper.Builder builder = new AggregateMetricFieldMapper.Builder(name);
+            TypeParsers.parseField(builder, name, node, parserContext);
```
Let's not call parseField, none of the properties that it parses are supported by this field?

csoulios replied:
Fixed

iverase commented on x-pack/plugin/rollup/src/main/java/org/elasticsearch/xpack/rollup/mapper/AggregateMetricFieldMapper.java:
```diff
@@ -0,0 +1,340 @@
+
+    @Override
+    protected void parseCreateField(ParseContext context, List<IndexableField> fields) throws IOException {
+        Object metricsAsObject;
```
I think this logic should in the parse() method not here? This method should never be called.

csoulios replied:
By checking code at [`FieldMapper#parse()` ](https://github.com/elastic/elasticsearch/blob/f7a5b739729bb00c221d0ad54f0c5062876640da/server/src/main/java/org/elasticsearch/index/mapper/FieldMapper.java#L277) it seems that `parse()` is calling `parseCreateField()` and only handles any exceptions thrown().  

Finally it calls `multiFields.parse(this, context);` which does nothing our case. I do not really understand why overriding `parse()` is preferable to `parseCreateField()`. Happy to change this though.

jimczi replied:
++ to use `parseCreateField`. We parse the content from the `_source` so I don't see why it should never be called @iverase ?

iverase replied:
I must say I am not sure but so far has always been my choice. I will check.

iverase commented on x-pack/plugin/rollup/src/main/java/org/elasticsearch/xpack/rollup/mapper/AggregateMetricFieldMapper.java:
```diff
@@ -0,0 +1,340 @@
+                            throw new IllegalArgumentException("Aggregate metric [" + metricName +
+                                "] of field [" + fieldType.name() + "] must be an integer number");
+                        } else {
+                            context.addIgnoredField(fieldType.name());
```
I don't think we should just ignore and go to the next entry of the map if ignore malformed is true. My take is that we should not generate the doc value and ignore this field.

iverase commented on x-pack/plugin/rollup/src/main/java/org/elasticsearch/xpack/rollup/mapper/AggregateMetricFieldMapper.java:
```diff
@@ -0,0 +1,340 @@
+            throw new IllegalArgumentException("Aggregate metric field [" + fieldType.name() +
+                "] must contain all metrics " + metrics.value().toString());
+        }
+
```
Not sure about this. If we do not have all the metrics and ignore_value is true, then we should not throw an error but we should not generate the doc value either?

iverase reviewed (changes requested):
Thanks @csoulios. 

I think the parse logic should be on the parse method not the parseCreateField method. The latter is used (I think) for dynamic mapping and I don't think this field allow that.

I am not sure the leniency in the parse logic. I would like to confirm it but if there is an error on one of the metrics (does not exist or have an incorrect value) we should stop generating the doc value if ignore malformed is true.

csoulios commented on x-pack/plugin/rollup/src/main/java/org/elasticsearch/xpack/rollup/mapper/AggregateMetricFieldMapper.java:
```diff
@@ -0,0 +1,340 @@
+
+        if (fieldType().hasDocValues()) {
+            BytesStreamOutput streamOutput = new BytesStreamOutput();
+            streamOutput.writeMap(parsedMetricsMap);
```
Question to @iverase and @jimczi : 

Is this method of serializing the `parsedMetricsMap` preferred as opposed to 
```
            for (Map.Entry<String, Object> e : parsedMetricsMap.entrySet()) {
                streamOutput.writeString(e.getKey());

                Double d = null;
                if (e.getValue() instanceof Double) {
                    d = (Double) e.getValue();
                } else if (e.getValue() instanceof Integer) {
                    d = Double.valueOf((Integer) e.getValue());
                }
                streamOutput.writeLong(Double.doubleToRawLongBits(d));

            }
````
We should take into consideration: 

- How it's best to deserialize this map.
- Values are now Double objects, but we may want to serialize histrograms or cardinality structures.

jimczi replied:
I think we said that we should give each metric his dedicated sub-field ? In such case you can create one docvalue field for each metric.

iverase replied:
I actually prefer this way and I would recommend to use `ByteBuffersDataOutput` from Lucene instead of BytesStreamOutput. The idea behind is that when deserialising you can use `ByteArrayDataInput` which it can be reused and you some allocations per document.

See Histogram field type for reference. wdyt?

csoulios committed:
Addressing code review comments

jimczi commented on x-pack/plugin/rollup/src/main/java/org/elasticsearch/xpack/rollup/mapper/AggregateMetricFieldMapper.java:
```diff
@@ -0,0 +1,340 @@
+        static {
+            FIELD_TYPE.setTokenized(false);
+            FIELD_TYPE.setHasDocValues(true);
+            FIELD_TYPE.setIndexOptions(IndexOptions.NONE);
```
I think we want to index by default but in such case we should add the `DoublePoint`s when parsing the document

jimczi commented on x-pack/plugin/rollup/src/main/java/org/elasticsearch/xpack/rollup/mapper/AggregateMetricFieldMapper.java:
```diff
@@ -0,0 +1,340 @@
+            return AggregateMetricFieldMapper.Defaults.IGNORE_MALFORMED;
+        }
+
+        public AggregateMetricFieldMapper.Builder metrics(Set<String> metrics) {
```
Should we use an `Enum` here to ensure that we restrict to metrics that we can handle (e.g.: `max`, `min`, `avg`) ?

csoulios replied:
Fixed

jimczi commented on x-pack/plugin/rollup/src/main/java/org/elasticsearch/xpack/rollup/mapper/AggregateMetricFieldMapper.java:
```diff
@@ -0,0 +1,340 @@
+
+    public static class Defaults {
+        public static final Explicit<Boolean> IGNORE_MALFORMED = new Explicit<>(false, false);
+        public static final Explicit<Set<String>> METRICS = new Explicit<>(Collections.emptySet(), false);
```
The default should contain all metrics ? Otherwise it's an empty field ?

jimczi commented on x-pack/plugin/rollup/src/main/java/org/elasticsearch/xpack/rollup/mapper/AggregateMetricFieldMapper.java:
```diff
@@ -0,0 +1,340 @@
+        }
+    }
+
+    public static final class AggregateMetricFieldType extends MappedFieldType {
```
I think you can use a plain `NumericFieldType` instance for each metric and delegate the execution based on the context ? For now this would all delegate to the default metric and we think at how metrics aggs should communicate a different choice as a follow up.

jimczi commented on x-pack/plugin/rollup/src/main/java/org/elasticsearch/xpack/rollup/mapper/AggregateMetricFieldMapper.java:
```diff
@@ -0,0 +1,340 @@
+
+    private Explicit<Set<String>> metrics;
+
+    private AggregateMetricFieldMapper(String simpleName, MappedFieldType fieldType, MappedFieldType defaultFieldType,
```
I think we should disable `MultiFields` and `CopyTo` for this field.

jimczi commented on x-pack/plugin/rollup/src/main/java/org/elasticsearch/xpack/rollup/mapper/AggregateMetricFieldMapper.java:
```diff
@@ -0,0 +1,340 @@
+        if (context.externalValueSet()) {
+            metricsAsObject = context.externalValue();
+        } else {
+            metricsAsObject = context.parser().map();
```
Can you try to use the parser instead to check that the current token is an Object and then iterate over the keys ?
If you stream the parsing you can then use a `NumberFieldMapper` instance for each metric and delegate the `parseCreateField` when needed.

jimczi commented on x-pack/plugin/rollup/src/main/java/org/elasticsearch/xpack/rollup/mapper/AggregateMetricFieldMapper.java:
```diff
@@ -0,0 +1,340 @@
+import java.util.stream.Collectors;
+
+/** A {@link FieldMapper} for a field containing aggregate metrics such as min/max/value_count etc. */
+public class AggregateMetricFieldMapper extends FieldMapper {
```
Should we call it `AggregateDoubleMetricFieldMapper` ?

jimczi reviewed (commented):
Thanks @csoulios . I think you can use a dedicated module for this field mapper (see `mapper-flattened`), no need to put it in the rollup module.
I left the other comments in the pr.

csoulios committed:
Moved AggregateMetricFieldMapper to its own module

polyfractal commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/XPackSettings.java:
```diff
@@ -134,6 +134,10 @@ private XPackSettings() {
     public static final Setting<Boolean> VECTORS_ENABLED = Setting.boolSetting("xpack.vectors.enabled", true, Setting.Property.NodeScope);
 
+    /** Setting for enabling or disabling aggregate_metric fields. Defaults to true. */
+    public static final Setting<Boolean> AGGREGATE_METRIC_ENABLED = Setting.boolSetting("xpack.aggregate_metric.enabled",
```
Since this is targeting Basic, I don't believe we need the `enabled` flag and associated code (e.g. that's what we did with the `analytics` plugin: https://github.com/elastic/elasticsearch/pull/46092/)

csoulios replied:
Fixed

polyfractal commented on x-pack/plugin/mapper-aggregate-metric/src/main/java/org/elasticsearch/xpack/aggregatemetric/mapper/AggregateMetricFieldMapper.java:
```diff
@@ -0,0 +1,339 @@
+                if (propName.equals(Names.METRICS.getPreferredName())) {
+                    String metricsStr[] = XContentMapValues.nodeStringArrayValue(propNode);
+                    // Convert the array of Metric enum to a Set of their string representation
+                    Set<String> supporteddMetrics = Arrays.stream(Metrics.values()).map(s -> s.toString()).collect(Collectors.toSet());
```
I wonder if we should use an `EnumSet` here rather than converting back to strings?  Or implementing something like [`fromString()` on the enum](https://github.com/elastic/elasticsearch/blob/master/server/src/main/java/org/elasticsearch/search/aggregations/bucket/histogram/DateIntervalWrapper.java#L68) to parse the string and return the enum (otherwise throw an exception).

Since we already have the enum it'd be nice if we could convert into enum space rather than other way around.

csoulios replied:
Fixed

polyfractal commented on x-pack/plugin/mapper-aggregate-metric/src/main/java/org/elasticsearch/xpack/aggregatemetric/mapper/AggregateMetricFieldMapper.java:
```diff
@@ -0,0 +1,339 @@
+                            parsedMetrics.add(metricsStr[i]);
+                        }
+                    }
+                    builder.metrics(parsedMetrics);
```
And if we do end up creating a set of enums, we'd need to change the builder (and all downstream users) to accept the enum instead of strings.

csoulios replied:
Fixed

polyfractal commented on x-pack/plugin/mapper-aggregate-metric/src/main/java/org/elasticsearch/xpack/aggregatemetric/mapper/AggregateMetricFieldMapper.java:
```diff
@@ -0,0 +1,339 @@
+            return;
+        }
+
+        Map<String, Object> parsedMetricsMap = new LinkedHashMap<>();
```
Do we need the insertion order maintained via `LinkedHashMap` instead of a regular `HashMap`?

csoulios replied:
Changed this to a `EnumMap<Metrics, Object>`

polyfractal commented on x-pack/plugin/mapper-aggregate-metric/src/main/java/org/elasticsearch/xpack/aggregatemetric/mapper/AggregateMetricFieldMapper.java:
```diff
@@ -0,0 +1,339 @@
+                    }
+                }
+
+                // Make sure that value_count is not a negative value
```
I think this comment is outdated... looks like the negative check is a bit further down

csoulios replied:
fixed

polyfractal commented on x-pack/plugin/mapper-aggregate-metric/src/main/java/org/elasticsearch/xpack/aggregatemetric/mapper/AggregateMetricFieldMapper.java:
```diff
@@ -0,0 +1,339 @@
+
+                // Make sure that value_count is not a negative value
+                if (Metrics.value_count.toString().equals(metricName) == true) {
+                    if (e.getValue() instanceof Integer == false) {
```
Explicitly looking for Integer makes me a little nervous, as we might get different types (and want to support up to a long anyhow I think?).  Perhaps we could reuse the `long` parsing from NumberFieldMapper?  E.g.

```
long metricValue = NumberFieldMapper.NumberType.LONG.parse(e.getValue(), false)
```

With coerce to false, that will reject floating points and strings, but correctly parse any other long/integer/short/byte/etc.

csoulios replied:
fixed

polyfractal commented on x-pack/plugin/mapper-aggregate-metric/src/main/java/org/elasticsearch/xpack/aggregatemetric/mapper/AggregateMetricFieldMapper.java:
```diff
@@ -0,0 +1,339 @@
+                "] must contain all metrics " + metrics.value().toString());
+        }
+
+        if (fieldType().hasDocValues()) {
```
I wonder if we should override `setHasDocValues()` on the MappedFieldType so that it throws an exception if the user tries to disable doc values?  Since we don't index anything and only store doc values, seems like we shouldn't let the user disable them?

Not sure, WDYT?

polyfractal reviewed (commented):
Left a few comments! :)

Open question for discussion: since this is isolated functionality at the moment (can configure but not actually use the fields yet), should we potentially merge to a feature branch instead?  Not sure of our plans to backport to 7x, but it would make a potential backport (or revert) easier too.

csoulios committed:
Fix broken doc test

csoulios committed:
Fixes to address code review comments

csoulios commented:
@elasticmachine run elasticsearch-ci/bwc
@elasticmachine run elasticsearch-ci/2

csoulios committed:
Merge branch 'master' into rollups

csoulios committed:
Fixes to address code review comments.

Make parsing of malformed values stricter and
ignore all field metrics even if a single metric
is wrong

csoulios committed:
Fix broken integration tests

csoulios commented:
@elasticmachine run elasticsearch-ci/bwc
@elasticmachine run elasticsearch-ci/default-distro

csoulios committed:
Added delegate field mappers of NumberFieldType

csoulios committed:
Merge branch 'master' into rollups

csoulios committed:
Delegate queries to NumberFieldType fields

csoulios committed:
Merge branch 'master' into rollups

csoulios committed:
Style fixes

jimczi commented on x-pack/plugin/mapper-aggregate-metric/src/main/java/org/elasticsearch/xpack/aggregatemetric/mapper/AggregateDoubleMetricFieldMapper.java:
```diff
@@ -0,0 +1,448 @@
+                delegateFieldMapper.parse(context);
+
+                if (Metric.value_count == metric) {
+                    Number n = context.doc().getField(delegateFieldMapper.simpleName()).numericValue();
```
```suggestion
                    Number n = context.doc().getField(delegateFieldMapper.fieldType().name())).numericValue();
```
You need the full name to access the field.

csoulios replied:
Fixed

jimczi commented on x-pack/plugin/mapper-aggregate-metric/src/test/java/org/elasticsearch/xpack/aggregatemetric/mapper/AggregateDoubleMetricTypeTests.java:
```diff
@@ -0,0 +1,152 @@
+
+import static org.elasticsearch.xpack.aggregatemetric.mapper.AggregateDoubleMetricFieldMapper.Names.METRICS;
+
+public class AggregateDoubleMetricTypeTests extends ESSingleNodeTestCase {
```
We usually write this kind of test for field mappers as unit tests. You could move them in AggregateDoubleMetricFieldTypeTests or in AggregateDoubleMetricFieldMapperTests and tests solely that the queries are correctly created (you don't need to index and search 1000 docs).

jimczi replied:
We also need a real integration test but only to verify that the mapper is correctly registered so a simple rest test should be enough. Can you add one in https://github.com/elastic/elasticsearch/tree/master/x-pack/plugin/src/test/resources/rest-api-spec/test ?

csoulios replied:
Fixed

jimczi commented on x-pack/plugin/rollup/src/main/java/org/elasticsearch/xpack/rollup/Rollup.java:
```diff
@@ -187,4 +187,5 @@ public void close() {
             schedulerEngine.get().stop();
         }
     }
+
```
nit: remove

csoulios replied:
Fixed

jimczi reviewed (commented):
It looks good @csoulios, I left some comments regarding the tests. Can you also create a branch to merge this pr (can be in your fork) since we said that we'll need metrics aggregation support to publish in master ?

csoulios committed:
Nit: removed blank line

csoulios committed:
Addressed reviewer comments

csoulios committed:
Fixed NPE issue when "metrics" field is missing

csoulios committed:
Added integration test

csoulios committed:
Merge branch 'master' into rollups

csoulios commented on x-pack/plugin/mapper-aggregate-metric/src/main/java/org/elasticsearch/xpack/aggregatemetric/mapper/AggregateDoubleMetricFieldMapper.java:
```diff
@@ -0,0 +1,448 @@
+            EnumMap<Metric, NumberFieldMapper> metricMappers = new EnumMap<>(Metric.class);
+            // Instantiate one NumberFieldMapper instance for each metric
+            for (Metric m : this.metrics) {
+                String fieldName = name + "._" + m.name();
```
Fixed

csoulios commented on x-pack/plugin/mapper-aggregate-metric/src/main/java/org/elasticsearch/xpack/aggregatemetric/mapper/AggregateDoubleMetricFieldMapper.java:
```diff
@@ -0,0 +1,448 @@
+        protected Explicit<Metric> defaultMetric(BuilderContext context) {
+            if (defaultMetric != null) {
+                if (metrics != null && metrics.contains(defaultMetric) == false) {
+                    // The default_metric is not defined in the the "metrics" field
```
Fixed

csoulios commented on x-pack/plugin/mapper-aggregate-metric/src/main/java/org/elasticsearch/xpack/aggregatemetric/mapper/AggregateDoubleMetricFieldMapper.java:
```diff
@@ -0,0 +1,398 @@
+                builder.store(false);
+                builder.index(true);
+                builder.docValues(true);
+                builder.fieldType().setDocValuesType(DocValuesType.NUMERIC);
```
Fixed

csoulios committed:
Override AggregateDoubleMetricFieldMapper methods

isFieldWithinQuery to handle range queries
efficiently,
fielddataBuilder to handle aggregations and sort,
docValueFormat to format double correctly

iverase commented on x-pack/plugin/mapper-aggregate-metric/src/main/java/org/elasticsearch/xpack/aggregatemetric/mapper/AggregateDoubleMetricFieldMapper.java:
```diff
@@ -0,0 +1,481 @@
+                    iterator.remove();
+                } else if (TypeParsers.parseMultiField(builder, name, parserContext, propName, propNode)) {
+                    iterator.remove();
+                }
```
What happen if any of the above conditions are met?

iverase commented on x-pack/plugin/mapper-aggregate-metric/src/main/java/org/elasticsearch/xpack/aggregatemetric/mapper/AggregateDoubleMetricFieldMapper.java:
```diff
@@ -0,0 +1,481 @@
+        if (context.externalValueSet() == true) {
+            throw new IllegalArgumentException("Field [" + name() + "] of type [" + typeName() + "] can't be used in multi-fields");
+        }
+
```
should we handle the case when the value is an array of metrics?

csoulios replied:
I cannot think of a case that an array aggregate metrics can be useful. Therefore, we don't plan to support it

iverase replied:
It is ok not to support it but we should capture that case and throw a meaningful error?

iverase replied:
See:   https://github.com/elastic/elasticsearch/blob/eade4f03f426447fea6148f7cf482ee0baee4cd3/x-pack/plugin/analytics/src/main/java/org/elasticsearch/xpack/analytics/mapper/HistogramFieldMapper.java#L413

iverase commented on x-pack/plugin/mapper-aggregate-metric/src/main/java/org/elasticsearch/xpack/aggregatemetric/mapper/AggregateDoubleMetricFieldMapper.java:
```diff
@@ -0,0 +1,481 @@
+    @Override
+    protected void doMerge(Mapper mergeWith) {
+        super.doMerge(mergeWith);
+        AggregateDoubleMetricFieldMapper other = (AggregateDoubleMetricFieldMapper) mergeWith;
```
Not sure but I would expect that merging two fields with different metrics should be an error.

iverase reviewed (commented):
I understand the idea now. It is looking good, just left a few comments for clarification.

csoulios committed:
Merge branch 'feature/aggregate-metrics' into rollups

csoulios committed:
Ensure that a metric field cannot be an array

csoulios committed:
Ensure that merging two fields with different

metrics should be an error.

csoulios committed:
Merge branch 'master' into rollups

csoulios committed:
Merge branch 'master' into rollups

csoulios committed:
Checkstyle

csoulios committed:
Fix typo

csoulios committed:
Merge branch 'feature/aggregate-metrics' into rollups

csoulios committed:
Applied "spotless"

iverase commented on x-pack/plugin/mapper-aggregate-metric/src/main/java/org/elasticsearch/xpack/aggregatemetric/mapper/AggregateDoubleMetricFieldMapper.java:
```diff
@@ -0,0 +1,543 @@
+                            + "] of type ["
+                            + typeName()
+                            + "] does not support indexing multiple values for the same field in the same document"
+                    );
```
nit: same field -> same metric?

csoulios committed:
Fixed broken doc-test after merge

iverase reviewed (commented):
Thanks @csoulios. Easy to follow the mapping logic. I would only recommend to add a test that uses a parent object in the mapping so we make sure we are resolving the path correctly in the metric's docValues in that case.

csoulios committed:
Added test for nested aggregate_metric field

csoulios committed:
Merge branch 'feature/aggregate-metrics' into rollups

csoulios commented:
@elasticmachine run elasticsearch-ci/1
@elasticmachine run elasticsearch-ci/packaging-sample-unix-docker
